File: sdk/typescript/src/utils/epoch.ts
import { EPOCH_LENGTH, WINDOWS_PER_EPOCH, WINDOW_LENGTH } from "../constants/chain.js";

export function computeEpoch(
  blockHeight: number,
  epochLength: number = EPOCH_LENGTH,
): number {
  return Math.floor((blockHeight - 1) / epochLength);
}

export function computeWindow(
  blockHeight: number,
  epochLength: number = EPOCH_LENGTH,
  windowsPerEpoch: number = WINDOWS_PER_EPOCH,
): number {
  const windowLength = Math.floor(epochLength / windowsPerEpoch);
  return Math.floor(((blockHeight - 1) % epochLength) / windowLength);
}

export function epochStartBlock(epochNumber: number, epochLength: number = EPOCH_LENGTH): number {
  return epochNumber * epochLength + 1;
}

export function epochEndBlock(epochNumber: number, epochLength: number = EPOCH_LENGTH): number {
  return (epochNumber + 1) * epochLength;
}

export function windowStartBlock(
  epochStartBlockHeight: number,
  windowIndex: number,
  windowLength: number = WINDOW_LENGTH,
): number {
  return epochStartBlockHeight + windowIndex * windowLength;
}

export function windowEndBlock(
  epochStartBlockHeight: number,
  windowIndex: number,
  windowLength: number = WINDOW_LENGTH,
): number {
  return epochStartBlockHeight + windowIndex * windowLength + windowLength - 1;
}

export function formatProgress(elapsed: number, total: number): string {
  return `${elapsed}/${total}`;
}

File: sdk/python/seocheon/utils/epoch.py
"""Epoch and window computation utilities for the Seocheon SDK."""


def compute_epoch(block_height: int, epoch_length: int) -> int:
    """Calculate the epoch number for a given block height.

    epoch_number = (block_height - 1) / epoch_length
    """
    if block_height <= 0 or epoch_length <= 0:
        return 0
    return (block_height - 1) // epoch_length


def compute_window(block_height: int, epoch_length: int, windows_per_epoch: int) -> int:
    """Calculate the window index within an epoch for a given block height.

    window_index = ((block_height - 1) % epoch_length) / window_length
    """
    if epoch_length <= 0 or windows_per_epoch <= 0:
        return 0
    window_length = epoch_length // windows_per_epoch
    return ((block_height - 1) % epoch_length) // window_length


def epoch_start_block(epoch_number: int, epoch_length: int) -> int:
    """Return the starting block height for the given epoch number."""
    return epoch_number * epoch_length + 1


def epoch_end_block(epoch_number: int, epoch_length: int) -> int:
    """Return the ending block height for the given epoch number."""
    return (epoch_number + 1) * epoch_length


def window_start_block(epoch_start: int, window_index: int, window_length: int) -> int:
    """Return the starting block height for a given window within an epoch."""
    return epoch_start + window_index * window_length


def window_end_block(window_start: int, window_length: int) -> int:
    """Return the ending block height for a given window."""
    return window_start + window_length - 1
